### PR TITLE
[SPARK-44341][SQL][PYTHON] Define the computing logic through PartitionEvaluator API and use it in WindowExec and WindowInPandasExec

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
`WindowExec` and `WindowInPandasExec` are updated to use the `PartitionEvaluator` API to do execution.

### Why are the changes needed?
To define the computing logic and requires the caller side to explicitly list what needs to be serialized and sent to executors

### Does this PR introduce _any_ user-facing change?
'No'.
Just update the inner implementation.

### How was this patch tested?
Add new test cases.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @cloud-fan @viirya  cc @vinodkc
The CI failure looks unrelated.

**Comment by beliefer** [comment]
cc @HyukjinKwon @zhengruifeng

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowEvaluatorFactory.scala [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+    }
+  }
+
+}
+
+class WindowEvaluatorFactory(
+    val windowExpression: Seq[NamedExpression],
+    val partitionSpec: Seq[Expression],
+    val orderSpec: Seq[SortOrder],
+    val conf: SQLConf,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowEvaluatorFactory.scala:298]: we usually don't need to pass the SQLConf, as worker side can get it via `SQLConf.get`

**Comment by cloud-fan** [comment]
We can probably skip testing it. Overall it's just a refactor and it's probably too much to run many tests twice. We can enable it by default later so that all tests cover this code path.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowEvaluatorFactory.scala [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+import org.apache.spark.sql.execution.metric.SQLMetric
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{CalendarIntervalType, DateType, DayTimeIntervalType, DecimalType, IntegerType, TimestampNTZType, TimestampType, YearMonthIntervalType}
+import org.apache.spark.util.collection.Utils
+
+trait WindowEvaluatorFactoryBase {
+  def windowExpression: Seq[NamedExpression]
+  def partitionSpec: Seq[Expression]
+  def orderSpec: Seq[SortOrder]
+  def conf: SQLConf
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowEvaluatorFactory.scala:37]: Seems `conf` is always `SQLConf.get` in both implementations?

**Review by viirya — COMMENTED** [review]
Refactored code looks okay to me.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowEvaluatorFactory.scala (hunk 2 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowEvaluatorFactory.scala:37]: Yes. I have used `SQLConf.get` in both implementations.

**Comment by viirya** [comment]
As tests are skipped now, maybe you can enable it to test change in CI, and revert back current config before merging.

**Comment by beliefer** [comment]
> As tests are skipped now, maybe you can enable it to test change in CI, and revert back current config before merging.

In fact, CI passed in previous commit.

**Comment by viirya** [comment]
> > As tests are skipped now, maybe you can enable it to test change in CI, and revert back current config before merging.
> 
> In fact, CI passed in previous commit.

Oh okay. 👍

**Review by viirya — APPROVED** [review]

**Comment by beliefer** [comment]
The CI failure is unrelated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/WindowInPandasEvaluatorFactory.scala [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+
+    def lowerBoundIndex(frameIndex: Int) = boundIndices(frameIndex)._1
+
+    def upperBoundIndex(frameIndex: Int) = boundIndices(frameIndex)._2
+
+    (requiredIndices.sum, lowerBoundIndex, upperBoundIndex, windowBoundTypes)
+  }
+
+  // Unwrap the expressions and factories from the map.
+  private val expressionsWithFrameIndex =
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/WindowInPandasEvaluatorFactory.scala:125]: so all these `private val`s will be serialized to the executor side, right? I think it's better to define them in the evaluator class, not the factor, to reduce the data to be serialized.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/WindowInPandasEvaluatorFactory.scala:125]: cc @Hisoka-X can you also check your PRs (either merged or not)? We should avoid putting variables in the factor class, but define them in the evaluator class instead.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan @viirya @Hisoka-X Thank you.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowEvaluatorFactory.scala [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Add, AggregateWindowFunction, Ascending, Attribute, BoundReference, CurrentRow, DateAdd, DateAddYMInterval, DecimalAddNoOverflowCheck, Descending, Expression, FrameLessOffsetWindowFunction, FrameType, IdentityProjection, IntegerLiteral, JoinedRow, MutableProjection, NamedExpression, OffsetWindowFunction, PythonFuncExpression, RangeFrame, RowFrame, RowOrdering, SortOrder, SpecificInternalRow, SpecifiedWindowFrame, TimeAdd, TimestampAddYMInterval, UnaryMinus, UnboundedFollowing, UnboundedPreceding, UnsafeProjection, UnsafeRow, WindowExpression}
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
+import org.apache.spark.sql.execution.ExternalAppendOnlyUnsafeRowArray
+import org.apache.spark.sql.execution.metric.SQLMetric
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{CalendarIntervalType, DateType, DayTimeIntervalType, DecimalType, IntegerType, TimestampNTZType, TimestampType, YearMonthIntervalType}
+import org.apache.spark.util.collection.Utils
+
+trait WindowEvaluatorFactoryBase {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowEvaluatorFactory.scala:33]: with hindsight, shall we move it to a new file? Its two sub-classes are not both in this file, so it's a bit weird to put it with one of the sub-class.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowEvaluatorFactory.scala:33]: OK. Let's move it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExec.scala [hunk 5]
```diff
@@ -95,111 +95,23 @@ case class WindowExec(
...
-            throw new NoSuchElementException
-          }
-        }
+    // Start processing.
+    if (conf.usePartitionEvaluator) {
+      child.execute().mapPartitionsWithEvaluator(evaluatorFactory)
+    } else {
+      child.execute().mapPartitions { iter =>
+        val evaluator = evaluatorFactory.createEvaluator()
+        evaluator.eval(0, iter)
```
Comment (vinodkc) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExec.scala:114]: @ beliefer, Can you please raise a follow-up PR to handle the partition index as this https://github.com/apache/spark/pull/42185

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExec.scala:114]: Thank you for the reminder. I will take a look!

**Pull request closed without merging**